### PR TITLE
spectral-extraction: improved initial parameter estimation

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request improves the automatic initial input parameter determination for input to spectral extraction, particularly for cases where the previous defaults would result in a window outside the range of the image which would throw an error from specreduce.  This also prevents other cases like this from throwing a traceback by instead raising an error snackbar message on load and pointing to the plugin for manual extraction.

**Re failing coverage checks**: I did add test coverage for the near-edge example to ensure it forces a one-sided background, but since the parser only accepts files, its difficult to add coverage for the "brightest pixel on edge" logic and the complete try/except snackbar.

**TODO**:
- [x] iterate on expected behavior
- [x] add test coverage

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Review by camipacifici — APPROVED** [review]
Looks good to me!

File: jdaviz/configs/specviz2d/helper.py [hunk 1]
```diff
@@ -149,14 +149,19 @@ def load_data(self, spectrum_2d=None, spectrum_1d=None, spectrum_1d_label=None,
...
-                _ = spext.create_extract(add_data=True)
-
-                # Warn that this shouldn't be used for science
-                msg = ("The extracted 1D spectrum was generated automatically."
-                       " See the spectral extraction plugin for details or to"
-                       " perform a custom extraction.")
-                msg = SnackbarMessage(msg, color='warning', sender=self,
-                                      timeout=10000)
+                try:
+                    _ = spext.create_extract(add_data=True)
```
Comment (pllim) [inline jdaviz/configs/specviz2d/helper.py:153]: Might as well clean it up a bit while you are at it?
```suggestion
                    spext.create_extract(add_data=True)
```

File: jdaviz/configs/specviz2d/helper.py [hunk 2]
```diff
@@ -149,14 +149,19 @@ def load_data(self, spectrum_2d=None, spectrum_1d=None, spectrum_1d_label=None,
...
+                except Exception:
+                    msg = ("Automatic spectrum extraction failed.  See the spectral extraction"
+                           " plugin to perform a custom extraction")
+                    msg = SnackbarMessage(msg, color='error', sender=self, timeout=10000)
```
Comment (pllim) [inline jdaviz/configs/specviz2d/helper.py:157]:
```suggestion
                    msg = SnackbarMessage(
                        "Automatic spectrum extraction failed.  See the spectral extraction"
                        " plugin to perform a custom extraction", color='error', sender=self, timeout=10000)
```

File: jdaviz/configs/specviz2d/helper.py [hunk 3]
```diff
@@ -149,14 +149,19 @@ def load_data(self, spectrum_2d=None, spectrum_1d=None, spectrum_1d_label=None,
...
+                else:
+                    # Warn that this shouldn't be used for science
+                    msg = ("The extracted 1D spectrum was generated automatically."
+                           " See the spectral extraction plugin for details or to"
+                           " perform a custom extraction.")
+                    msg = SnackbarMessage(msg, color='warning', sender=self,
+                                          timeout=10000)
```
Comment (pllim) [inline jdaviz/configs/specviz2d/helper.py:164]:
```suggestion
                    msg = SnackbarMessage(
                        "The extracted 1D spectrum was generated automatically."
                        " See the spectral extraction plugin for details or to"
                        " perform a custom extraction.", color='warning', sender=self, timeout=10000)
```

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py [hunk 4]
```diff
@@ -237,7 +252,7 @@ def marks(self):
 
         viewer2d = self.app.get_viewer('spectrum-2d-viewer')
         viewer1d = self.app.get_viewer('spectrum-viewer')
-        if not viewer2d.state.reference_data or not viewer1d.state.reference_data:
+        if not viewer2d.state.reference_data:
```
Comment (pllim) [inline jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:255]: What does the removal of the second condition do?

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/tests/test_spectral_extraction.py [hunk 5]
```diff
@@ -72,3 +72,14 @@ def test_plugin(specviz2d_helper):
...
+
+@pytest.mark.remote_data
+def test_spectrum_on_top(specviz2d_helper):
+    fn = download_file('https://mast.stsci.edu/api/v0.1/Download/file/?uri=mast:jwst/product/jw01529-c1002_t002_miri_p750l_s2d.fits', cache=True)  # noqa
+
+    specviz2d_helper.load_data(spectrum_2d=fn)
+
+    pext = specviz2d_helper.app.get_tray_item_from_name('spectral-extraction')
+    assert pext.bg_type_selected == 'OneSided'
+    assert pext.bg_separation < 0
```
Comment (pllim) [inline jdaviz/configs/specviz2d/plugins/spectral_extraction/tests/test_spectral_extraction.py:85]: What is `< 0` testing? You sure you don't want to be more explicit with `numpy.testing.assert_allclose()`?

**Review by pllim — COMMENTED** [review]
Is there a follow-up issue to make Specviz2D parser more flexible and then add the edge case you couldn't add before in this PR?

**Comment by kecnry** [comment]
> Is there a follow-up issue to make Specviz2D parser more flexible

Yes there is, I'll add a note in there to also extend the test cases [:cat:](https://jira.stsci.edu/browse/JDAT-2671)

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py (hunk 4 above)
Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:255]: previously this avoided tracebacks before the original 1d spectrum was extracted.  Now this will force that error to occur, which is now caught by the try/except and exposed in the snackbar.

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/tests/test_spectral_extraction.py (hunk 5 above)
Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/tests/test_spectral_extraction.py:85]: It's really just testing that the logic detected that the extraction window was so close to the top of the image, that the background region needed to be below the extraction region.  I was intentionally non-specific in the test, because we'll very likely improve the separation/width default logic in the future which would change the exact value, but as long as the window-detection logic doesn't change significantly to avoid the edge entirely, that should still always return a negative separation.

**Review by pllim — APPROVED** [review]
LGTM. Thanks!
